{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prerequisites:\n",
    "- Pandas\n",
    "- glob\n",
    "- subprocess\n",
    "\n",
    "#### Introduction\n",
    "This script guides you through the process of firstly, extending the initial Risk SNPs list (seed snps, PD or AD.) with Haploreg (Linkage disequilibrium) and GTex (eQTLs), and secondly, map those snps to their corresponding genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Query Haploreg (https://pubs.broadinstitute.org/mammals/haploreg/haploreg.php) v4.1 with the initial list of seed snps provides in the folder for both PD and AD separately. They are called \"PD_SNPS.txt\" for Parkinson's Disease or \"AD_SNPS.txt\" for Alzheimers's Disease.\n",
    "\n",
    "    - options: default, set output mode to Text, LD Threshold 0.8 \n",
    "    \n",
    "###### Important: Haploreg is starting to get slow when snps count exceeds ~ 80. So we first divide the the seed list into equal snps of a modarate size. Then we use it as input separately for Haploreg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the path to your input file = newline separated list of seed snps\n",
    "# in this example we are using PD snps. Change the path according to your study (AD or PD)\n",
    "seed_snps=\"AD_SNPs_meemansa.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the initial snps list and divide into smaller subsets. \n",
    "#Those subsets are getting saved in the current working directory, called \"snps_subset_xx\".\n",
    "snps_per_file = 80\n",
    "smallfile=None\n",
    "snps_list=[] #list of seed snps for gtex later\n",
    "with open(seed_snps) as snps: # thi is the newline separated file of seed snps for PD. CHange this file according to your inpt of choice above.\n",
    "    for lineno, line in enumerate(snps):\n",
    "        snps_list.append(line.rstrip())\n",
    "        if lineno % snps_per_file == 0:\n",
    "            small_filename = 'snps_subset_{}.txt'.format(lineno + snps_per_file)\n",
    "            smallfile = open(small_filename,\"w\")\n",
    "        smallfile.write(line)\n",
    "    if smallfile:\n",
    "        smallfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " #### Now query Haploreg manually on their website with the smaller files separately and put them into the current working directory\n",
    " #### Follow these steps\n",
    " 1. go to (https://pubs.broadinstitute.org/mammals/haploreg/haploreg.php\n",
    " 2. build query\n",
    " 3. build query with each subset, output format: Text, LD Threshold: 0.8\n",
    " 4. Save results as \"haploreg_x\" with x = 1,2,3,4 etc. in the current working directory. Important to name them in that convention"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#list of all files we created\n",
    "files=glob.glob(\"haploreg*\")\n",
    "\n",
    "#read all files dymanically and merge them into a big dataframe\n",
    "haploreg_df=pd.read_csv(files[0],delimiter = '\\t')\n",
    "for file in files[1:]:\n",
    "    current_df=pd.read_csv(file,delimiter = '\\t')\n",
    "    haploreg_df = haploreg_df.append(current_df) #the final big dataframe contaiing all snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos_hg38</th>\n",
       "      <th>r2</th>\n",
       "      <th>D'</th>\n",
       "      <th>is_query_snp</th>\n",
       "      <th>rsID</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>AFR</th>\n",
       "      <th>AMR</th>\n",
       "      <th>...</th>\n",
       "      <th>GENCODE_id</th>\n",
       "      <th>GENCODE_name</th>\n",
       "      <th>GENCODE_direction</th>\n",
       "      <th>GENCODE_distance</th>\n",
       "      <th>RefSeq_id</th>\n",
       "      <th>RefSeq_name</th>\n",
       "      <th>RefSeq_direction</th>\n",
       "      <th>RefSeq_distance</th>\n",
       "      <th>dbSNP_functional_annotation</th>\n",
       "      <th>query_snp_rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>33974719</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rs35833468</td>\n",
       "      <td>TAAC</td>\n",
       "      <td>T</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.52</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000129521.8</td>\n",
       "      <td>EGLN3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NM_022073</td>\n",
       "      <td>EGLN3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>23640.0</td>\n",
       "      <td>.</td>\n",
       "      <td>rs35833468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>9756586</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rs3219012</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000114026.16</td>\n",
       "      <td>OGG1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NM_016828</td>\n",
       "      <td>OGG1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NSM;INT</td>\n",
       "      <td>rs3219012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>171588461</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2421847</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.02</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000117523.11</td>\n",
       "      <td>PRRC2C</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NM_015172</td>\n",
       "      <td>PRRC2C</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NSM</td>\n",
       "      <td>rs2421847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>50699212</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0</td>\n",
       "      <td>rs12592161</td>\n",
       "      <td>G</td>\n",
       "      <td>C,T</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.3</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000138600.5</td>\n",
       "      <td>SPPL2A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8096.0</td>\n",
       "      <td>NM_032802</td>\n",
       "      <td>SPPL2A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8326.0</td>\n",
       "      <td>.</td>\n",
       "      <td>rs8035452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15</td>\n",
       "      <td>50699437</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0</td>\n",
       "      <td>rs12910731</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.3</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000138600.5</td>\n",
       "      <td>SPPL2A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7871.0</td>\n",
       "      <td>NM_032802</td>\n",
       "      <td>SPPL2A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8101.0</td>\n",
       "      <td>.</td>\n",
       "      <td>rs8035452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1400</th>\n",
       "      <td>5</td>\n",
       "      <td>8.89102e+07</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>rs10085009</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.08</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000248309.1</td>\n",
       "      <td>CTC-454M9.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NM_001131005</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6073.0</td>\n",
       "      <td>.</td>\n",
       "      <td>rs9293506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401</th>\n",
       "      <td>5</td>\n",
       "      <td>8.89126e+07</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>rs10056732</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.08</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000248309.1</td>\n",
       "      <td>CTC-454M9.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NM_001131005</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8537.0</td>\n",
       "      <td>.</td>\n",
       "      <td>rs9293506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1402</th>\n",
       "      <td>5</td>\n",
       "      <td>8.89138e+07</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>rs28662206</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.08</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000248309.1</td>\n",
       "      <td>CTC-454M9.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NM_001131005</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9737.0</td>\n",
       "      <td>.</td>\n",
       "      <td>rs9293506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1403</th>\n",
       "      <td>5</td>\n",
       "      <td>8.89148e+07</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>rs13357047</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.08</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000248309.1</td>\n",
       "      <td>CTC-454M9.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NM_001131005</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10732.0</td>\n",
       "      <td>.</td>\n",
       "      <td>rs9293506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1404</th>\n",
       "      <td>19</td>\n",
       "      <td>4.49178e+07</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rs3925681</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.35</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000130208.4</td>\n",
       "      <td>APOC1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NM_001645</td>\n",
       "      <td>APOC1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INT</td>\n",
       "      <td>rs3925681</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57756 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr     pos_hg38    r2    D' is_query_snp        rsID   ref  alt    AFR  \\\n",
       "0     14     33974719     1     1            1  rs35833468  TAAC    T   0.35   \n",
       "1      3      9756586     1     1            1   rs3219012     C    T    NaN   \n",
       "2      1    171588461     1     1            1   rs2421847     A    G  0.002   \n",
       "3     15     50699212  0.91  0.96            0  rs12592161     G  C,T   0.19   \n",
       "4     15     50699437  0.89  0.95            0  rs12910731     A    G   0.19   \n",
       "...   ..          ...   ...   ...          ...         ...   ...  ...    ...   \n",
       "1400   5  8.89102e+07     1     1            0  rs10085009     T    A   0.23   \n",
       "1401   5  8.89126e+07     1     1            0  rs10056732     C    T   0.23   \n",
       "1402   5  8.89138e+07     1     1            0  rs28662206     A    G   0.23   \n",
       "1403   5  8.89148e+07     1     1            0  rs13357047     G    A   0.23   \n",
       "1404  19  4.49178e+07     1     1            1   rs3925681     G    A   0.22   \n",
       "\n",
       "       AMR  ...          GENCODE_id  GENCODE_name GENCODE_direction  \\\n",
       "0     0.52  ...   ENSG00000129521.8         EGLN3               0.0   \n",
       "1      NaN  ...  ENSG00000114026.16          OGG1               0.0   \n",
       "2     0.02  ...  ENSG00000117523.11        PRRC2C               0.0   \n",
       "3      0.3  ...   ENSG00000138600.5        SPPL2A               3.0   \n",
       "4      0.3  ...   ENSG00000138600.5        SPPL2A               3.0   \n",
       "...    ...  ...                 ...           ...               ...   \n",
       "1400  0.08  ...   ENSG00000248309.1   CTC-454M9.1               0.0   \n",
       "1401  0.08  ...   ENSG00000248309.1   CTC-454M9.1               0.0   \n",
       "1402  0.08  ...   ENSG00000248309.1   CTC-454M9.1               0.0   \n",
       "1403  0.08  ...   ENSG00000248309.1   CTC-454M9.1               0.0   \n",
       "1404  0.35  ...   ENSG00000130208.4         APOC1               0.0   \n",
       "\n",
       "     GENCODE_distance     RefSeq_id RefSeq_name RefSeq_direction  \\\n",
       "0                 0.0     NM_022073       EGLN3              5.0   \n",
       "1                 0.0     NM_016828        OGG1              0.0   \n",
       "2                 0.0     NM_015172      PRRC2C              0.0   \n",
       "3              8096.0     NM_032802      SPPL2A              3.0   \n",
       "4              7871.0     NM_032802      SPPL2A              3.0   \n",
       "...               ...           ...         ...              ...   \n",
       "1400              0.0  NM_001131005       MEF2C              5.0   \n",
       "1401              0.0  NM_001131005       MEF2C              5.0   \n",
       "1402              0.0  NM_001131005       MEF2C              5.0   \n",
       "1403              0.0  NM_001131005       MEF2C              5.0   \n",
       "1404              0.0     NM_001645       APOC1              0.0   \n",
       "\n",
       "     RefSeq_distance dbSNP_functional_annotation query_snp_rsid  \n",
       "0            23640.0                           .     rs35833468  \n",
       "1                0.0                     NSM;INT      rs3219012  \n",
       "2                0.0                         NSM      rs2421847  \n",
       "3             8326.0                           .      rs8035452  \n",
       "4             8101.0                           .      rs8035452  \n",
       "...              ...                         ...            ...  \n",
       "1400          6073.0                           .      rs9293506  \n",
       "1401          8537.0                           .      rs9293506  \n",
       "1402          9737.0                           .      rs9293506  \n",
       "1403         10732.0                           .      rs9293506  \n",
       "1404             0.0                         INT      rs3925681  \n",
       "\n",
       "[57756 rows x 33 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#test\n",
    "haploreg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "##keep column 6 and 25 (index 5 and 24). \"SNP ID\" and \"Gene name\"\n",
    "haploreg_df=haploreg_df[['rsID','GENCODE_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NaN values removed\n"
     ]
    }
   ],
   "source": [
    "#remove NaN valued rows and shady values \n",
    "haploreg_df_unique=haploreg_df.dropna(inplace=False)\n",
    "print(\"NaN values removed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " #### Query Haploreg is done.\n",
    " ### 2.\n",
    " #### Now we we query gtex portal (http://gtexportal.org/home/index.html) to extend our mapping with eQTLs mapping. Since both Haploreg and gtex portal use genome build hg38, there's no further genome coordinate mapping needed.\n",
    " \n",
    " #### We access it through the API and fetch only Brain Tissue eQTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('snps_gene_eqtls.csv', 'a') as outfile: # save in csv in current working directory\n",
    "    for snp in snps_list:\n",
    "        #fetch brain eQTLs only\n",
    "        command=\"curl -X GET --header \\'Accept: text/html\\' \\'https://gtexportal.org/rest/v1/association/singleTissueEqtl?format=tsv&snpId=\"+snp+\"&tissueSiteDetailId=Brain_Amygdala%2CBrain_Anterior_cingulate_cortex_BA24%2CBrain_Caudate_basal_ganglia%2CBrain_Cerebellar_Hemisphere%2CBrain_Cerebellum%2CBrain_Cortex%2CBrain_Frontal_Cortex_BA9%2CBrain_Hippocampus%2CBrain_Hypothalamus%2CBrain_Nucleus_accumbens_basal_ganglia%2CBrain_Putamen_basal_ganglia%2CBrain_Spinal_cord_cervical_c-1%2CBrain_Substantia_nigra&datasetId=gtex_v8\\'\"\n",
    "        subprocess.call(command,shell=True,stdout=outfile) #that might take a few minutes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           snpId  geneSymbolUpper\n",
      "0      rs1858446             GDI2\n",
      "1      rs1858446             GDI2\n",
      "2      rs1858446             GDI2\n",
      "3      rs1858446             GDI2\n",
      "4      rs1858446             GDI2\n",
      "...          ...              ...\n",
      "10006      snpId  geneSymbolUpper\n",
      "10007      snpId  geneSymbolUpper\n",
      "10008  rs1138272           DOC2GP\n",
      "10009  rs1138272           DOC2GP\n",
      "10010      snpId  geneSymbolUpper\n",
      "\n",
      "[10011 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "#open the output\n",
    "eQTL_query=pd.read_csv(\"snps_gene_eqtls.csv\",sep=\"\\t\")\n",
    "#keep only snpID and gene symbol column\n",
    "eQTL_query=eQTL_query[[\"snpId\",\"geneSymbolUpper\"]]\n",
    "print (eQTL_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size without duplicates: 1882\n"
     ]
    }
   ],
   "source": [
    "#drop duplicates\n",
    "eQTL_query=eQTL_query.drop_duplicates(keep=\"first\",inplace=False)\n",
    "print (\"Size without duplicates:\",len(eQTL_query))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snpId</th>\n",
       "      <th>geneSymbolUpper</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1858446</td>\n",
       "      <td>GDI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>snpId</td>\n",
       "      <td>geneSymbolUpper</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs59043219</td>\n",
       "      <td>IRF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>ZNF276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>MC1R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>VPS9D1-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>SPIRE2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>VPS9D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>URAHP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>RP11-368I7.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         snpId  geneSymbolUpper\n",
       "0    rs1858446             GDI2\n",
       "5        snpId  geneSymbolUpper\n",
       "15  rs59043219             IRF6\n",
       "22   rs2228479           ZNF276\n",
       "23   rs2228479             MC1R\n",
       "25   rs2228479       VPS9D1-AS1\n",
       "27   rs2228479           SPIRE2\n",
       "30   rs2228479           VPS9D1\n",
       "32   rs2228479            URAHP\n",
       "43   rs2228479     RP11-368I7.6"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eQTL_query.head(10) #row number 6 contains column headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final size 1881\n",
      "         snpId geneSymbolUpper\n",
      "0    rs1858446            GDI2\n",
      "15  rs59043219            IRF6\n",
      "22   rs2228479          ZNF276\n",
      "23   rs2228479            MC1R\n",
      "25   rs2228479      VPS9D1-AS1\n",
      "27   rs2228479          SPIRE2\n",
      "30   rs2228479          VPS9D1\n",
      "32   rs2228479           URAHP\n",
      "43   rs2228479    RP11-368I7.6\n",
      "56  rs10217950             CHM\n"
     ]
    }
   ],
   "source": [
    "#we drop 2th row because it contains column headers. \n",
    "#Since we removed all duplicates, this row is the only instance of column headers within the data.\n",
    "eQTL_query=eQTL_query.drop(eQTL_query.index[1]) #drop it\n",
    "print(\"final size\",len(eQTL_query))\n",
    "print (eQTL_query.head(10)) # done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snpId</th>\n",
       "      <th>geneSymbolUpper</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1858446</td>\n",
       "      <td>GDI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs59043219</td>\n",
       "      <td>IRF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>ZNF276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>MC1R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>VPS9D1-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9977</th>\n",
       "      <td>rs600491</td>\n",
       "      <td>RP11-67L3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9984</th>\n",
       "      <td>rs10410539</td>\n",
       "      <td>AC006277.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9994</th>\n",
       "      <td>rs11142</td>\n",
       "      <td>PSRC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>rs11142</td>\n",
       "      <td>SYPL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10008</th>\n",
       "      <td>rs1138272</td>\n",
       "      <td>DOC2GP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1881 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            snpId geneSymbolUpper\n",
       "0       rs1858446            GDI2\n",
       "15     rs59043219            IRF6\n",
       "22      rs2228479          ZNF276\n",
       "23      rs2228479            MC1R\n",
       "25      rs2228479      VPS9D1-AS1\n",
       "...           ...             ...\n",
       "9977     rs600491     RP11-67L3.4\n",
       "9984   rs10410539      AC006277.2\n",
       "9994      rs11142           PSRC1\n",
       "9995      rs11142           SYPL2\n",
       "10008   rs1138272          DOC2GP\n",
       "\n",
       "[1881 rows x 2 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eQTL_query"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. \n",
    "#### Now combine Haploreg and GTex results.\n",
    "####  Important: only consider eQTL SNPs, which are in Haploreg results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Drop snps from the GTex reults that are NOT in Haploreg results\n",
    "haploreg_ids=haploreg_df_unique.rsID.tolist() #haploreg snps ids in list\n",
    "eQTL_query_haplofilter=eQTL_query[eQTL_query[\"snpId\"].isin(haploreg_ids)] #filter Gtex reults using the haploreg ID list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#renaming eQTL_query_haplofilter column names to make it consistent with the Haploreg dataframe\n",
    "eQTL_query_haplofilter.columns=[\"rsID\",\"GENCODE_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>rsID</th>\n",
       "      <th>GENCODE_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">haplo</th>\n",
       "      <th>0</th>\n",
       "      <td>rs35833468</td>\n",
       "      <td>EGLN3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs3219012</td>\n",
       "      <td>OGG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs2421847</td>\n",
       "      <td>PRRC2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs12592161</td>\n",
       "      <td>SPPL2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs12910731</td>\n",
       "      <td>SPPL2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">gtex</th>\n",
       "      <th>9977</th>\n",
       "      <td>rs600491</td>\n",
       "      <td>RP11-67L3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9984</th>\n",
       "      <td>rs10410539</td>\n",
       "      <td>AC006277.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9994</th>\n",
       "      <td>rs11142</td>\n",
       "      <td>PSRC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>rs11142</td>\n",
       "      <td>SYPL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10008</th>\n",
       "      <td>rs1138272</td>\n",
       "      <td>DOC2GP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59280 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   rsID GENCODE_name\n",
       "haplo 0      rs35833468        EGLN3\n",
       "      1       rs3219012         OGG1\n",
       "      2       rs2421847       PRRC2C\n",
       "      3      rs12592161       SPPL2A\n",
       "      4      rs12910731       SPPL2A\n",
       "...                 ...          ...\n",
       "gtex  9977     rs600491  RP11-67L3.4\n",
       "      9984   rs10410539   AC006277.2\n",
       "      9994      rs11142        PSRC1\n",
       "      9995      rs11142        SYPL2\n",
       "      10008   rs1138272       DOC2GP\n",
       "\n",
       "[59280 rows x 2 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# finally join both Haploreg_final and eQTL_query_haplofilter \n",
    "frames = [haploreg_df_unique,eQTL_query_haplofilter]\n",
    "snps_gene_haploreg_gtex = pd.concat(frames, keys=['haplo', 'gtex'])\n",
    "#\n",
    "snps_gene_haploreg_gtex #we are almost done... ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of the final Dataframe: 37477\n"
     ]
    }
   ],
   "source": [
    "#remove duplicates from joined dataframe\n",
    "snps_gene_haploreg_gtex_unique=snps_gene_haploreg_gtex.drop_duplicates(keep=\"first\",inplace=False)\n",
    "print(\"Size of the final Dataframe:\",len(snps_gene_haploreg_gtex_unique)) #done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the final dataframe to your current working directory\n",
    "snps_gene_haploreg_gtex.to_csv(\"snps_gene_haploreg_gtex_meemansa_final.csv\",index=False,header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
